[Reshiro] iP

Reshiro opened the pull request:
# YoChan
> Ough... - Yo-Chan

YoChan tries her best to manage your tasks for you! It's
- text-based
- easy to learn
- ~~very~~ *extremely* fast to use.

All you need to do is
1. download it from [here](https://github.com/Reshiro/ip/releases/)
2. double-click it
3. add your tasks
4. let Yo-Chan remember and manage your tasks for you!

As Yo-Chan is a seal, she does it all for **free**! 👍

Features:
- [X] Managing tasks
- [X] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class YoChan {
    public static void main(String[] args) {
        new YoChan("data", "YoChan.txt").run();
    }
}

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Reshiro committed:
Change chatbot name and greeting

Reshiro committed:
Add support for echo

Reshiro committed:
Add support for tasks

Reshiro committed:
Add marking and unmarking of tasks

Reshiro committed:
Add todo, deadline, and event task types

Reshiro committed:
Add testing of current commands

Reshiro committed:
Add error handling for invalid inputs

Reshiro committed:
Add support for deletion of tasks

Reshiro committed:
Add saving of tasks to disk

Reshiro committed:
Add loading of tasks from disk

Reshiro committed:
Merge branch 'branch-Level-7' into master

Reshiro committed:
Replace date and time strings with LocalDateTime

Reshiro committed:
Fix coding style violations

Reshiro committed:
Merge branch 'branch-Level-8' into master

Reshiro committed:
Split functionality of driver class into multiple classes

Reshiro committed:
Fix coding style and javadoc issues

Reshiro committed:
Add comments to explain command parsing

Reshiro committed:
Merge branch 'branch-A-MoreOOP' into master

Reshiro committed:
Divide classes into packages

Reshiro committed:
Merge branch 'branch-A-Packages' into master

Reshiro committed:
Merge 'add-gradle-support' into master

Reshiro committed:
Fix task loading bug caused by refactoring

Reshiro committed:
Add Gradle support for build automation

Reshiro committed:
Add JUnit tests for the Parser and Ui classes

Reshiro committed:
Add JUnit tests for the TaskList class

Reshiro committed:
Merge branch 'branch-A-JUnit' into master

Reshiro committed:
Add JavaDoc comments for public methods

Reshiro committed:
Mark some fields as final to enforce immutability

Reshiro committed:
Add implementation of Find command

Reshiro committed:
Merge branch 'branch-A-JavaDoc' into master

Reshiro committed:
Merge branch 'branch-A-CodingStandard' into master

Reshiro committed:
Merge branch 'branch-Level-9' into master

Reshiro committed:
Add more comments to the Parser class

Reshiro committed:
Merge branch 'branch-A-CheckStyle'

Reshiro committed:
Extract Find command parsing into a method

Reshiro committed:
Merge branch 'branch-A-CheckStyle' into master

Reshiro committed:
Fix README format

LeeJiaWei23 commented on src/test/java/yochan/ParserTest.java:
```diff
@@ -0,0 +1,82 @@
+import yochan.task.Event;
+import yochan.task.Todo;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
```
Should these static import be ordered at the top of the file instead?

LeeJiaWei23 commented on src/test/java/yochan/TaskListTest.java:
```diff
@@ -0,0 +1,82 @@
+ * @author Michael Cheong (Reshiro)
+ */
+public class TaskListTest {
+    @Test
+    public void markTask_todoMark_success() {
+        TaskList taskList = new TaskList();
+        taskList.add(new Todo("test"));
+        try {
+            taskList.markTask(0);
+        } catch (Exception e) {
+            fail();
+        }
+        Todo todoTest = new Todo("test");
+        todoTest.mark();
+        assertEquals(taskList.get(0).toString().trim(),
+                todoTest.toString().trim());
+    }
+
+    @Test
+    public void markTask_invalidIndex_exceptionThrown() {
+        TaskList taskList = new TaskList();
+        try {
+            taskList.markTask(0);
+        } catch (Exception e) {
+            assertEquals(e.getMessage(),
+                    "Invalid task number! >:(");
+        }
+    }
```
Perhaps you could consider adding JavaDoc here as well.

LeeJiaWei23 reviewed (commented):
LGTM. Just some small details. 👍

GCheeYang commented on src/main/java/yochan/TaskList.java:
```diff
@@ -0,0 +1,80 @@
+     *
+     * @throws YoChanException If the index is invalid.
+     */
+    public void markTask(int index) throws YoChanException {
+        if (isValidIndex(index)) {
+            tasks.get(index).mark();
+        } else {
+            throw new YoChanException("Invalid task number! >:(");
+        }
+    }
+
+    /**
+     * Marks the indicated task as incomplete.
+     *
+     * @throws YoChanException If the index is invalid.
+     */
+    public void unmarkTask(int index) throws YoChanException {
+        if (isValidIndex(index)) {
+            tasks.get(index).unmark();
+        } else {
+            throw new YoChanException("Invalid task number! >:(");
+        }
+    }
```
Perhaps this could be changed to a toggle function instead of marking and unmarking to make the code more seamless?

GCheeYang commented on src/main/java/yochan/Ui.java:
```diff
@@ -0,0 +1,150 @@
+     *
+     * @param task The task that was marked.
+     */
+    public void showTaskMarked(Task task) {
+        showBorder();
+        System.out.println("Ough! I've marked this task as done:");
+        System.out.println(task);
+        showBorder();
+    }
+
+    /**
+     * Prints the String representation of the task that was unmarked.
+     *
+     * @param task The task that was unmarked.
+     */
+    public void showTaskUnmarked(Task task) {
+        showBorder();
+        System.out.println("Ough... I've marked this task as not done yet:");
+        System.out.println(task);
+        showBorder();
+    }
```
The string representation sounds like it is changing it from marked to unmarked and vice versa, maybe could be changed to "this task is marked as..." ?

GCheeYang commented on src/main/java/yochan/command/MarkCommand.java:
```diff
@@ -0,0 +1,29 @@
+    /**
+     * Creates a MarkCommand object with the index of the Task to be marked.
+     */
+    public MarkCommand(int taskNumber) {
+        this.taskNumber = taskNumber;
+    }
```
Maybe change the class name to MarkTaskCommand? So that it is clearer what the command is doing

GCheeYang commented on src/main/java/yochan/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @author Michael Cheong (Reshiro)
+ */
+public class UnmarkCommand extends Command {
+    private final int taskNumber;
+
```
Maybe change the name to UnmarkTaskCommand so that is is clearer

Reshiro committed:
Add basic GUI functionality

Reshiro committed:
Add command functionality to the GUI

Reshiro committed:
Merge branch 'branch-Level-10' into master

Reshiro committed:
Fix ExitCommand execution for the GUI

Reshiro committed:
Fix ExitCommand execution for the GUI

Reshiro committed:
Merge branch 'branch-Level-10' into master

Reshiro committed:
Add assertions to document assumptions that should hold

Reshiro committed:
Improve code quality in various methods.

The parseCommand method is too long and should be refactored.

Splitting the behavior of this method into other methods and calling
these methods from parseCommand would allow the code to be more
readable.

Creating a method for commands that do not need to be specially
parsed and a method for those that do allows for greater separation
of responsibilities.

Reshiro committed:
Merge pull request #2 from Reshiro/branch-A-Assertions

The program currently functions based on certain assumptions such as the ability for the chatbot to return a non-null response.

Assert statements should be used to check for these assumptions to ensure that the program is working as intended.

Reshiro committed:
Merge branch 'master' into branch-A-CodeQuality

Reshiro committed:
Merge pull request #3 from Reshiro/branch-A-CodeQuality

Long methods that perform multiple distinct responsibilities are present, such as parseCommand in the Parser class.

Methods such as parseCommand can instead call helper functions that operate at a certain level of abstraction.

Splitting methods in such a way allows the program to adhere to the Single Layer of Abstraction Principle.

Reshiro committed:
Add the priority command to indicate the priority of a task

Tasks in the task list can only be differentiated based on their
description.

The user may have trouble identifiying their most important tasks
without additional context.

Let's create a priority command to assign priorities to tasks.

A command that can be used on created tasks maintains the interface for
other commands.

Reshiro committed:
Merge branch 'branch-BCD-Extension' into master

Reshiro committed:
Fix bugs related to negative priorities

Reshiro committed:
Move saved task parsing from Storage to Parser

Reshiro committed:
Fix coding style issues related to import order

Reshiro committed:
Remove mentions of Duke from the code

Reshiro committed:
Fix text-ui-test expected output

Reshiro committed:
Fix JUnit test output format and code quality

Reshiro committed:
Add chatbot name to the titlebar

Reshiro committed:
Add resizing support for the GUI

Reshiro committed:
Remove unnecessary css files

Reshiro committed:
Add image of the chatbot GUI

Reshiro committed:
Add a user guide for the chatbot

Reshiro committed:
Fix user guide formatting issues

Reshiro committed:
Add more details to the user guide

Reshiro committed:
Add user guide to repo README

Reshiro committed:
Update format of the user guide

Reshiro committed:
Add details to the user guide

Reshiro committed:
Add additional details to the user guide

Reshiro committed:
Fix user guide markdown formatting

Reshiro committed:
Fix date formatting issues
